tests/functional/states/test_kubernetes.py

```diff
@@ -0,0 +1,1426 @@
+    """
+    name = random_string("deployment-", uppercase=False)
+    namespace = "default"
+
+    # Only create the deployment if requested
+    if request.param:
+        ret = kubernetes.deployment_present(
+            name=name,
+            namespace=namespace,
+            spec=_deployment_spec,
+            wait=True,
+        )
+        assert ret.result is True
+    try:
+        yield {"name": name, "namespace": namespace}
+    finally:
+        ret = kubernetes.deployment_absent(name=name, namespace=namespace, wait=True)
+        assert ret.result is True
+        assert ret.comment in ["The deployment does not exist", "None"]
+
+
+@pytest.mark.parametrize("deployment", [False], indirect=True)
+def test_deployment_present(kubernetes, deployment, _deployment_spec):
+    """
+    Test kubernetes.deployment_present creates a deployment
+    """
+    ret = kubernetes.deployment_present(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        spec=_deployment_spec,
+        wait=True,
+    )
+
+    assert ret.result is True
+    assert ret.changes
+
+    # Comment out idempotent test for now
+    # TODO: The state module needs fixed to handle proper present functionality
+
+
+# def test_deployment_present_idempotency(kubernetes, deployment, _deployment_spec):
+#     """
+#     Test kubernetes.deployment_present is idempotent
+#     """
+#     ret = kubernetes.deployment_present(
+#         name=deployment["name"],
+#         namespace=deployment["namespace"],
+#         spec=_deployment_spec,
+#         wait=True,
+#     )
+#     assert ret.result is True
+#     assert "already exists" in ret.comment
+#     assert not ret.changes
+
+
+def test_deployment_present_replace(kubernetes, deployment):
+    """
+    Test kubernetes.deployment_present replaces a deployment
+    """
+    new_spec = {
+        "replicas": 3,
+        "selector": {"matchLabels": {"app": "test"}},
+        "template": {
+            "metadata": {"labels": {"app": "test"}},
+            "spec": {
+                "containers": [
+                    {
+                        "name": "nginx",
+                        "image": "nginx:latest",
+                        "ports": [{"containerPort": 80}],
+                    }
+                ]
+            },
+        },
+    }
+
+    ret = kubernetes.deployment_present(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        spec=new_spec,
+        wait=True,
+    )
+
+    assert ret.result is True
+    assert ret.changes["spec"]["replicas"] == 3
+
+
+def test_deployment_present_test_mode(kubernetes, deployment, _deployment_spec):
+    """
+    Test kubernetes.deployment_present in test mode
+    """
+    ret = kubernetes.deployment_present(
+        name=deployment["name"],
+        namespace=deployment["namespace"],
+        spec=_deployment_spec,
+        test=True,
+    )
+
+    assert ret.result is None
+    assert not ret.changes
```

Comment:
suggestion: The deployment is present already. Since the module is not idempotent, afaict it always reports changes in non-test mode. This should be the same for test mode. See suggestions above about combining tests for both modes.